components/constellation/constellation.rs

```diff
@@ -264,18 +264,13 @@ pub struct Constellation<Message, LTF, STF> {
     /// WebRender thread.
     webrender_api_sender: webrender_api::RenderApiSender,
 
-    /// The set of all event loops in the browser. We generate a new
-    /// event loop for each registered domain name (aka eTLD+1) in
-    /// each top-level browsing context. We store the event loops in a map
-    /// indexed by top-level browsing context id
-    /// (as a `TopLevelBrowsingContextId`) and registered
-    /// domain name (as a `Host`) to event loops. This double
-    /// indirection ensures that separate tabs do not share event
-    /// loops, even if the same domain is loaded in each.
+    /// The set of all event loops in the browser.
+    /// We store the event loops in a map
+    /// indexed by registered domain name (as a `Host`) to event loops.
     /// It is important that scripts with the same eTLD+1
     /// share an event loop, since they can use `document.domain`
     /// to become same-origin, at which point they can share DOM objects.
-    event_loops: HashMap<TopLevelBrowsingContextId, HashMap<Host, Weak<EventLoop>>>,
+    event_loops: HashMap<Host, Weak<EventLoop>>,
```

Comment:
I don't understand this change. What happens if 2 tabs are from the same domain name now?

Reply:
With this change, if 2 tabs are from the same domain name, they will share an event-loop/script-thread. Previously, such sharing could only happen between same-domain browsing-contexts within one tab.

Reply:
But 2 tabs from the same domain name *shouldn't* share an event loop if they are unrelated, AFAIK.

Reply:
From what I understand, @asajeffrey suggested to do it this way for now to support `opener`, and to revisit it later as part of process isolation. See https://github.com/servo/servo/pull/20678#issuecomment-394759996

Reply:
@nox the problem is that adding opener complicates the definition of Unit Of Related Similar-Origin Browsing Contexts. Before adding opener, BCs in different tabs were unrelated, but with it BCs are related if they can be reached by opener. Sigh.

This is all currently in flux, e.g. https://groups.google.com/a/chromium.org/forum/m/#!topic/isolation-policy/zueJF9ad20g so I suspect the easiest thing to do is revisit this issue as part of process isolation.

Reply:
I've looked a bit into the spec, and would like to share the below findings and interpretations(which might be wrongly interpreted off course):

1. It is very clear from the spec that a [unit-of-related-similar-origin-browsing-contexts](https://html.spec.whatwg.org/multipage/browsers.html#unit-of-related-similar-origin-browsing-contexts)(who came up with that name!) **must share an event-loop**.
2. I cannot find a suggestion that same-origin tabs, that would not actually be part of such a unit, as **not allowed to share an event-loop**.
3. The difference between being part of such a unit or not, would be whether they can [directly reach](https://html.spec.whatwg.org/multipage/browsers.html#directly-reachable-browsing-contexts) each other, for example via the `opener`. So running on the same event-loop will not make them directly reachable in itself.
4. Also, such two same-origin tabs that aren't reachable will still be [familiar](https://html.spec.whatwg.org/multipage/browsers.html#familiar-with) with each other, and if somehow they actually knew each other's name, they would be [allowed to navigate](https://html.spec.whatwg.org/multipage/browsers.html#allowed-to-navigate) each other, see Step 6 of [#the-rules-for-choosing-a-browsing-context-given-a-browsing-context-name](https://html.spec.whatwg.org/multipage/browsers.html#the-rules-for-choosing-a-browsing-context-given-a-browsing-context-name) which states "there exists a browsing context whose name is the same as name, and current is familiar with that browsing context, and the user agent determines that the two browsing contexts are related enough that it is ok if they reach each other(...)". Also see in the big table above that the entry "name that exists with different top, if familiar and one permitted sandboxed navigator"...

What I conclude, is that it might actually be ok to run two same-origin tabs on the same event-loop, and we could also let them navigate each other by name(however this is separate from having a direct reference to each other via `opener`, or the return value of `window.open`).

Reply:
Opening two tabs with `github.com`, those two tabs should never share the same event loop.

A "unit of related similar-origin browsing contexts" is first and foremost a "unit of related browsing contexts", and 2 separate tabs are not a unit of related browsing contexts, because their browsing contexts aren't directly reachable from one another.

You are correct that they will be familiar, but I don't see why they would be allowed to navigate each other. At the very least this goes against what I've always had as a mental image of tabs and browsing contexts.

To me allowing unrelated tabs to share an event loop is too much of a Web regression to r+ that, so I'll defer to @jdm for the final decision.

Reply:
Ok. Another option we have, is to revert to some of the earlier patch I made with would, when creating the openee, check for a matching host within the event-loops of the opener only. It might not cover all edge cases, but it could be a start.

Reply:
@jdm, in case you missed the mention above in the barrage of messages, we(me, @nox and @asajeffrey) would like to ask you to settle the above discussion regarding a policy to share event loops. All the info you need is probably found above, and then some, however here is a summary:

1. Until this PR, Servo supported sharing an event-loop within a [unit-of-related-similar-origin-browsing-contexts](https://html.spec.whatwg.org/multipage/browsers.html#unit-of-related-similar-origin-browsing-contexts) only if all browsing contexts within that unit were all descendants of the same top-level-browsing-context. A top-level-browsing-context would act as a boundary outside of which sharing an event-loop wasn't possible.
2. This PR implements creating [auxiliary browsing contexts](https://html.spec.whatwg.org/multipage/#auxiliary-browsing-contexts), which are top-level-browsing-context that, if they have the same origin as their opener, would be part of the same "unit-of-related-similar-origin-browsing-contexts" as their opener, hence would need to share the same event-loop. This requires expanding the current policy. 
3. @asajeffrey suggested we simply share event-loops based on same-origin, which is the current implementation in this PR(see above). By definition this would make an opener/openee with the same origin share an event-loop. We would then revisit this sharing as part of upcoming work on process isolation.
4. @nox has reservations about the fact that this new sharing policy would also allow two unrelated same-origin tabs(which wouldn't be part of the same "unit-of-related-similar-origin-browsing-contexts") to share an event-loop.
5. From reading the spec only, I am of the opinion that it would seem more important for spec compliance to have one event-loop per "unit-of-related-similar-origin-browsing-contexts", than it is to prevent sharing of an event-loop with unrelated browsing contexts. If my reading is correct, the current proposed change would appear like a net win, and it has a lot of simplicity going for it. We could go beyond the spec and do something more complicated later to restrict sharing across tabs to openee/opener only, whereas by not going forward with the current change, we wouldn't have a spec compliant implementation of `opener`.

Reply:
I think the current implementation is fine, and we can prevent unrelated-but-same-origin browsing contexts from sharing event loops separately from this change. Please at least file an issue to keep track of that work.

Reply:
Thanks for looking into this, I've filed https://github.com/servo/servo/issues/21206

Reply:
@jdm Not sure what you mean, do you mean it's ok for now to have unrelated-but-same-origin browsing contexts share the same event loop?